File: consumers/src/lib/notify_consumers.cpp
#pragma once

#include <iostream>
#include <map>
#include <cppkafka/cppkafka.h>
#include <cpr/cpr.h>

// #include <opentelemetry/sdk/version/version.h>
#include <opentelemetry/trace/provider.h>

void ListenToKafkaTopicsAndNotifyTheConsumers(std::map<std::string, std::vector<std::string>> consumers, std::string kafka_brokers)
{
    std::cout << "KAFKA_BROKERS: " << kafka_brokers << std::endl;
    std::vector<std::string> topics;
    // get the topics from the map where all the keys are the topics that we need to listen to.
    for (auto it = consumers.begin(); it != consumers.end(); it++)
    {
        topics.push_back(it->first);
    }

    std::cout << "Listening to below topics... " << std::endl;
    for (const auto &topic : topics)
    {
        std::cout << topic << std::endl;
    }

    cppkafka::Configuration config = {
        {"metadata.broker.list", kafka_brokers}, {"group.id", "global_consumer"}, {"enable.auto.commit", false}};

    cppkafka::Consumer consumer(config);
    consumer.subscribe({topics});

    // Print the assigned partitions on assignment
    consumer.set_assignment_callback([](const cppkafka::TopicPartitionList &partitions)
                                     { std::cout << "Got assigned: " << partitions << std::endl; });

    // Print the revoked partitions on revocation
    consumer.set_revocation_callback([](const cppkafka::TopicPartitionList &partitions)
                                     { std::cout << "Got revoked: " << partitions << std::endl; });

    // check for messages in those topics

    // get the global tracer
    auto provider = opentelemetry::trace::Provider::GetTracerProvider();
    auto tracer = provider->GetTracer(std::getenv("SERVICE_NAME"));
    while (true)
    {
        cppkafka::Message message = consumer.poll();

        if (!message)
        {
            continue;
        }

        auto scoped_span = opentelemetry::trace::Scope(tracer->StartSpan(std::getenv("SERVICE_NAME")));
        if (message.get_error())
        {
            // librdkafka provides an error indicating we've reached the
            // end of a partition every time we do so. Make sure it's not one
            // of those cases, as it's not really an error
            if (!message.is_eof())
            {
                std::cout << "[+] Received error notification on topic -> " << message.get_topic() << " : " << message.get_error() << std::endl;
            }
            continue;
        }

        std::cout << "[+] Received message with key " << message.get_key() << " , topic -> " << message.get_topic() << " on Partition "
                  << message.get_partition() << ", offset " << message.get_offset() << ", with payload -> " << message.get_payload() << std::endl;

        // cppkafka::HeaderList headers = message.get_header_list() << std::endl;

        // std::ostream myout(std::cout.rdbuf());

        // std::ostringstream s2(std::ostringstream() << message.get_header_list());
        // myout << s2.str() << '\n';

        std::string topic = message.get_topic();
        std::vector<std::string> hosts = consumers[topic];

        for (int i = 0; i < hosts.size(); i++)
        {
            std::cout << "[+] Sending data to " << hosts[i] << " ..." << std::endl;
            cpr::Response response = cpr::Post(cpr::Url{hosts[i]},
                                               cpr::Body{std::string(message.get_payload())},
                                               cpr::Header{{"Content-Type", "application/json"}});
            if (response.status_code == 200)
            {
                std::cout << "✔️  Data Sent Successfully !!!" << std::endl;
                // Now commit the message
                consumer.commit(message);
            }
        }
    }
}

File: consumer-node/src/lib/notifyConsumers.js
import kafkajs from "kafkajs";
import { SpanStatusCode, trace } from "@opentelemetry/api";

const { Kafka } = kafkajs;
const serviceName = process.env["SERVICE_NAME"];

export async function ListenToKafkaTopicsAndNotifyTheConsumers(
  groupId,
  consumers,
  kafka_brokers
) {
  const kafka = new Kafka({
    clientId: serviceName,
    brokers: [kafka_brokers],
  });

  const consumer = kafka.consumer({ groupId });
  try {
    await consumer.connect();

    const topics = Object.keys(consumers);
    console.log(`[+] Listening for topics ${topics}`);

    await consumer.subscribe({ topics, fromBeginning: false });

    const tracer = trace.getTracer(serviceName);
    await consumer.run({
      eachMessage: async ({ topic, partition, message }) => {
        const { headers, value, key, offset } = message;
        let traceparent;
        Object.keys(headers).forEach((key) => {
          if (key.toLocaleLowerCase() === "traceparent") {
            traceparent = headers[key].toString();
          }
        });
        const payload = value.toString();
        console.log(
          `[+] Received message with key  ${key}, topic -> ${topic} on Partition ${partition}, offset ${offset} , with payload  ${payload}`
        );

        tracer.startActiveSpan(`${serviceName}`, async (span) => {
          span.setAttribute("key", key);
          span.setAttribute("partition", partition);
          span.setAttribute("offset", offset);
          span.setAttribute("topic", topic);

          try {
            await sendDataToConsumer(consumers[topic], payload, traceparent);
            console.log("✔️  Data Sent Successfully !!!");
            span.addEvent("log", {
              severity: "info",
              message: `Successfully exported data to ${consumers[topic]}.`,
            });
          } catch (err) {
            console.log("Error Sending Data !!!", err);
            span.recordException(err);
            span.setStatus({ code: SpanStatusCode.ERROR });
            span.addEvent("log", {
              severity: "critical",
              message: `unable to send data to ${consumers[topic]}.`,
            });
          } finally {
            span.end();
          }
        });
      },
    });

    ["SIGINT", "SIGTERM"].forEach((signal) => {
      process.on(signal, () => consumer.disconnect().catch(console.error));
    });
  } catch (err) {
    await consumer.disconnect();
    throw err;
  }
}

async function sendDataToConsumer(host, payload, traceparent) {
  console.log(`[+] Sending data to ${host}  ...`);
  const response = await fetch(`http://${host}`, {
    method: "POST",
    body: payload,
    headers: {
      "Content-type": "application/json",
      traceparent: traceparent,
    },
  });
  if (response.status !== 200) {
    throw Error(`cannot send data to ${host}`);
  }
}
